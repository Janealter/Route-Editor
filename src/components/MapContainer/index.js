import React, {Component} from 'react';
import {YMaps, GeoObject, Placemark} from 'react-yandex-maps';
import MapResponsive from '../MapResponsive';
import 'bootstrap/dist/css/bootstrap.css';
import PropTypes from "prop-types";

class MapContainer extends Component {
  constructor(props) {
    super(props);
    this.state = {
      mapInitialState: this.props.mapInitialState
    };
  }

  componentWillReceiveProps(nextProps) {
    if (nextProps.mapInitialState.center.join(',') !== this.state.mapInitialState.center.join(',') ||
      nextProps.mapInitialState.zoom !== this.state.mapInitialState.zoom) {
      this.setState({mapInitialState: nextProps.mapInitialState});
    }
  }

  render() {
    const placemarks = this.props.waypoints.map(waypoint =>
      <Placemark key={waypoint.id}
                 geometry={{coordinates: waypoint.coordinates}}
                 options={{
                   draggable: true,
                   iconLayout: 'islands#icon',
                   iconColor: '#007bff',
                   iconImageSize: [34, 40],
                   iconImageOffset: [-3, -42],
                   zIndex: 1
                 }}
                 properties={{
                   waypointId: waypoint.id,
                   balloonContent: waypoint.name
                 }}
                 onGeometryChange={this.props.onPlacemarkGeometryChange}
      />);
    return (
      <div className="card">
        <div className="card-header">
          Map
        </div>
        <div className="card-body">
          <div className="card">
            <YMaps>
              <MapResponsive state={this.state.mapInitialState}
                             height={450}
                             instanceRef={this.props.onMapNodeCreate}
                             responsive
              >
                {placemarks}
                <GeoObject geometry={{type: 'LineString', coordinates: this.props.waypoints.map(waypoint => waypoint.coordinates)}}
                          options={{
                            zIndex: 0,
                            strokeColor: '#dc3545',
                            strokeWidth: 3
                          }}
                />
              </MapResponsive>
            </YMaps>
          </div>
        </div>
      </div>
    );
  }
}

MapContainer.propTypes = {
  mapInitialState: PropTypes.object.isRequired,
  waypoints: PropTypes.array,
  onMapNodeCreate: PropTypes.func,
  onPlacemarkGeometryChange: PropTypes.func
};

export default MapContainer;
