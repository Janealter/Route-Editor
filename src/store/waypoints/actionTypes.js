const WAYPOINT_FETCHED = 'waypoints.WAYPOINT_FETCHED';
const WAYPOINT_ADDED = 'waypoints.WAYPOINT_ADDED';
const WAYPOINT_DELETED = 'waypoints.WAYPOINT_DELETED';
const WAYPOINT_MOVED = 'waypoints.WAYPOINT_MOVED';
const WAYPOINT_COORDINATES_CHANGED = 'waypoints.WAYPOINT_COORDINATES_CHANGED';

export { WAYPOINT_FETCHED, WAYPOINT_ADDED, WAYPOINT_DELETED, WAYPOINT_MOVED, WAYPOINT_COORDINATES_CHANGED };
