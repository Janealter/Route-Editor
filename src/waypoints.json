[
  {
    "id": 1,
    "name": "waypoint 1 very very very very very very large text here",
    "coordinates": [55.751678, 37.573346]
  },
  {
    "id": 2,
    "name": "waypoint 2",
    "coordinates": [55.768237, 37.5931097]
  },
  {
    "id": 3,
    "name": "waypoint 3 very large text here",
    "coordinates": [55.770913456, 37.5135454]
  },
  {
    "id": 4,
    "name": "waypoint 4 very very very very very very very very very very very very very very very very very very very very very large text",
    "coordinates": [55.7811245, 37.4930676]
  },
  {
    "id": 5,
    "name": "waypoint 5 very very very very very very very large text here",
    "coordinates": [55.792354, 37.6348678]
  },
  {
    "id": 6,
    "name": "waypoint 6 very large text here",
    "coordinates": [55.802354, 37.8348678]
  },
  {
    "id": 7,
    "name": "waypoint 7 very large text here",
    "coordinates": [55.902354, 37.7348678]
  }
]