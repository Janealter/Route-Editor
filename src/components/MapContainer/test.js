import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import MapContainer from "./index";
import waypoints from '../../waypoints.json';

describe('MapContainer', function () {
  const mapInitialState = { center: [50, 50], zoom: 10 };
  const onMapNodeCreateSpy = jest.fn();
  const onPlacemarkGeometryChangeSpy = jest.fn();
  const renderedComponent = shallow(
    <MapContainer mapInitialState={mapInitialState}
                  waypoints={waypoints}
                  onMapNodeCreate={onMapNodeCreateSpy}
                  onPlacemarkGeometryChange={onPlacemarkGeometryChangeSpy}
    />
  );

  it('renders self', () => {
    expect(renderedComponent.instance()).toBeInstanceOf(MapContainer);
  });

  it('changes state when receiving changed props', () => {
    renderedComponent.setProps({mapInitialState: {center: [45, 45], zoom: 8}});
    expect(renderedComponent.state().mapInitialState).toEqual({center: [45, 45], zoom: 8});
  });

  it('not changes state when receiving not changed props', () => {
    renderedComponent.setProps({mapInitialState: mapInitialState});
    expect(renderedComponent.state().mapInitialState).toEqual(mapInitialState);
  });

  it('the snapshot and the rendered component are the same', () => {
    expect(toJson(renderedComponent)).toMatchSnapshot();
  });
});