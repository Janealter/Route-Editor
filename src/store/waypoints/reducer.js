import Immutable from 'seamless-immutable';

import * as types from './actionTypes';

const initialState = Immutable({
  waypointsArray: []
});

const reduce = (state = initialState, action = {}) => {
  switch (action.type) {
    case types.WAYPOINT_FETCHED:
      return state.merge({
        waypointsArray: action.waypointsArray
      });
    case types.WAYPOINT_ADDED:
      return state.merge({
        waypointsArray: action.waypointsArray
      });
    case types.WAYPOINT_DELETED:
      return state.merge({
        waypointsArray: action.waypointsArray
      });
    case types.WAYPOINT_MOVED:
      return state.merge({
        waypointsArray: action.waypointsArray
      });
    case types.WAYPOINT_COORDINATES_CHANGED:
      return state.merge({
        waypointsArray: action.waypointsArray
      });
    default:
      return state;
  }
};

/**
 * Селекторы
 * */

const getWaypointsArray = state => {
  return state.waypoints.waypointsArray;
};

const getWaypointsByID = state => {
  const waypointsByID = {};

  state.waypoints.waypointsArray.forEach(waypoint => {
    waypointsByID[waypoint.id] = {
      name: waypoint.name,
      coordinates: waypoint.coordinates
    };
  });
  return waypointsByID;
};

const getWaypointsIDArray = state => {
  return state.waypoints.waypointsArray.map(waypoint => waypoint.id);
};

const getWaypointsNameArray = state => {
  return state.waypoints.waypointsArray.map(waypoint => waypoint.name);
};

export default reduce;

export { getWaypointsArray, getWaypointsByID, getWaypointsIDArray, getWaypointsNameArray };
