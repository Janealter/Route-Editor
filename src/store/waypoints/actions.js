import * as types from './actionTypes';
import backendService from '../../services/backend';
import * as waypointSelectors from './reducer';

const fetchWaypoints = () => {
  return (dispatch, getState) => {
    try {
      const waypointsArray = backendService.getWaypoints();
      dispatch({
        type: types.WAYPOINT_FETCHED,
        waypointsArray
      });
    } catch (e) {
      console.error(e);
    }
  }
};

const addWaypoint = (name, coordinates) => {
  return (dispatch, getState) => {
    try {
      const newWaypoint = {
        id: undefined,
        name: name,
        coordinates: coordinates
      };
      const id = backendService.addWaypoint(newWaypoint);
      const waypointsArray = waypointSelectors.getWaypointsArray(getState());
      const newWaypointsArray = [...waypointsArray];

      newWaypoint.id = id;
      newWaypointsArray.push(newWaypoint);

      dispatch({
        type: types.WAYPOINT_ADDED,
        waypointsArray: newWaypointsArray
      });
    } catch (e) {
      console.error(e);
    }
  }
};

const deleteWaypoint = (id) => {
  return (dispatch, getState) => {
    try {
      const waypointsArray = waypointSelectors.getWaypointsArray(getState());
      let newWaypointsArray = [...waypointsArray];

      newWaypointsArray = newWaypointsArray.filter(waypoint => waypoint.id !== id);

      dispatch({
        type: types.WAYPOINT_DELETED,
        waypointsArray: newWaypointsArray
      });
    } catch (e) {
      console.error(e);
    }
  }
};

const moveWaypoint = (oldIndex, newIndex) => {
  return (dispatch, getState) => {
    try {
      const waypointsArray = waypointSelectors.getWaypointsArray(getState());
      const newWaypointsArray = [...waypointsArray];
      const movedWaypoint = {};

      Object.assign(movedWaypoint, newWaypointsArray[oldIndex]);
      newWaypointsArray.splice(oldIndex, 1);
      newWaypointsArray.splice(newIndex, 0, movedWaypoint);

      dispatch({
        type: types.WAYPOINT_MOVED,
        waypointsArray: newWaypointsArray
      });
    } catch (e) {
      console.error(e);
    }
  }
};

const changeWaypointCoordinates = (id, coordinates) => {
  return (dispatch, getState) => {
    try {
      const waypointsArray = waypointSelectors.getWaypointsArray(getState());
      const newWaypointsArray = [...waypointsArray];
      const changedWaypoint = {};

      newWaypointsArray.some((waypoint, index) => {
        if (waypoint.id === id) {
          changedWaypoint.id = waypoint.id;
          changedWaypoint.name = waypoint.name;
          changedWaypoint.coordinates = coordinates;
          newWaypointsArray[index] = changedWaypoint;
          return true;
        }
        return false;
      });

      dispatch({
        type: types.WAYPOINT_COORDINATES_CHANGED,
        waypointsArray: newWaypointsArray
      });
    } catch (e) {
      console.error(e);
    }
  }
};

export { fetchWaypoints, addWaypoint, deleteWaypoint, moveWaypoint, changeWaypointCoordinates };
