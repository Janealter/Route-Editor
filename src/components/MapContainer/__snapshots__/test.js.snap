// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MapContainer the snapshot and the rendered component are the same 1`] = `
<div
  className="card"
>
  <div
    className="card-header"
  >
    Map
  </div>
  <div
    className="card-body"
  >
    <div
      className="card"
    >
      <YMaps
        enterprise={false}
        onApiAvaliable={[Function]}
        version="2.1"
      >
        <ResponsiveMap
          height={450}
          instanceRef={[MockFunction]}
          responsive={true}
          state={
            Object {
              "center": Array [
                50,
                50,
              ],
              "zoom": 10,
            }
          }
          width={400}
        >
          <Placemark
            geometry={
              Object {
                "coordinates": Array [
                  55.751678,
                  37.573346,
                ],
              }
            }
            key="1"
            onGeometryChange={[MockFunction]}
            options={
              Object {
                "draggable": true,
                "iconColor": "#007bff",
                "iconImageOffset": Array [
                  -3,
                  -42,
                ],
                "iconImageSize": Array [
                  34,
                  40,
                ],
                "iconLayout": "islands#icon",
                "zIndex": 1,
              }
            }
            properties={
              Object {
                "balloonContent": "waypoint 1 very very very very very very large text here",
                "waypointId": 1,
              }
            }
          />
          <Placemark
            geometry={
              Object {
                "coordinates": Array [
                  55.768237,
                  37.5931097,
                ],
              }
            }
            key="2"
            onGeometryChange={[MockFunction]}
            options={
              Object {
                "draggable": true,
                "iconColor": "#007bff",
                "iconImageOffset": Array [
                  -3,
                  -42,
                ],
                "iconImageSize": Array [
                  34,
                  40,
                ],
                "iconLayout": "islands#icon",
                "zIndex": 1,
              }
            }
            properties={
              Object {
                "balloonContent": "waypoint 2",
                "waypointId": 2,
              }
            }
          />
          <Placemark
            geometry={
              Object {
                "coordinates": Array [
                  55.770913456,
                  37.5135454,
                ],
              }
            }
            key="3"
            onGeometryChange={[MockFunction]}
            options={
              Object {
                "draggable": true,
                "iconColor": "#007bff",
                "iconImageOffset": Array [
                  -3,
                  -42,
                ],
                "iconImageSize": Array [
                  34,
                  40,
                ],
                "iconLayout": "islands#icon",
                "zIndex": 1,
              }
            }
            properties={
              Object {
                "balloonContent": "waypoint 3 very large text here",
                "waypointId": 3,
              }
            }
          />
          <Placemark
            geometry={
              Object {
                "coordinates": Array [
                  55.7811245,
                  37.4930676,
                ],
              }
            }
            key="4"
            onGeometryChange={[MockFunction]}
            options={
              Object {
                "draggable": true,
                "iconColor": "#007bff",
                "iconImageOffset": Array [
                  -3,
                  -42,
                ],
                "iconImageSize": Array [
                  34,
                  40,
                ],
                "iconLayout": "islands#icon",
                "zIndex": 1,
              }
            }
            properties={
              Object {
                "balloonContent": "waypoint 4 very very very very very very very very very very very very very very very very very very very very very large text",
                "waypointId": 4,
              }
            }
          />
          <Placemark
            geometry={
              Object {
                "coordinates": Array [
                  55.792354,
                  37.6348678,
                ],
              }
            }
            key="5"
            onGeometryChange={[MockFunction]}
            options={
              Object {
                "draggable": true,
                "iconColor": "#007bff",
                "iconImageOffset": Array [
                  -3,
                  -42,
                ],
                "iconImageSize": Array [
                  34,
                  40,
                ],
                "iconLayout": "islands#icon",
                "zIndex": 1,
              }
            }
            properties={
              Object {
                "balloonContent": "waypoint 5 very very very very very very very large text here",
                "waypointId": 5,
              }
            }
          />
          <Placemark
            geometry={
              Object {
                "coordinates": Array [
                  55.802354,
                  37.8348678,
                ],
              }
            }
            key="6"
            onGeometryChange={[MockFunction]}
            options={
              Object {
                "draggable": true,
                "iconColor": "#007bff",
                "iconImageOffset": Array [
                  -3,
                  -42,
                ],
                "iconImageSize": Array [
                  34,
                  40,
                ],
                "iconLayout": "islands#icon",
                "zIndex": 1,
              }
            }
            properties={
              Object {
                "balloonContent": "waypoint 6 very large text here",
                "waypointId": 6,
              }
            }
          />
          <Placemark
            geometry={
              Object {
                "coordinates": Array [
                  55.902354,
                  37.7348678,
                ],
              }
            }
            key="7"
            onGeometryChange={[MockFunction]}
            options={
              Object {
                "draggable": true,
                "iconColor": "#007bff",
                "iconImageOffset": Array [
                  -3,
                  -42,
                ],
                "iconImageSize": Array [
                  34,
                  40,
                ],
                "iconLayout": "islands#icon",
                "zIndex": 1,
              }
            }
            properties={
              Object {
                "balloonContent": "waypoint 7 very large text here",
                "waypointId": 7,
              }
            }
          />
          <GeoObject
            geometry={
              Object {
                "coordinates": Array [
                  Array [
                    55.751678,
                    37.573346,
                  ],
                  Array [
                    55.768237,
                    37.5931097,
                  ],
                  Array [
                    55.770913456,
                    37.5135454,
                  ],
                  Array [
                    55.7811245,
                    37.4930676,
                  ],
                  Array [
                    55.792354,
                    37.6348678,
                  ],
                  Array [
                    55.802354,
                    37.8348678,
                  ],
                  Array [
                    55.902354,
                    37.7348678,
                  ],
                ],
                "type": "LineString",
              }
            }
            instanceRef={[Function]}
            options={
              Object {
                "strokeColor": "#dc3545",
                "strokeWidth": 3,
                "zIndex": 0,
              }
            }
          />
        </ResponsiveMap>
      </YMaps>
    </div>
  </div>
</div>
`;
